device.go

```diff
@@ -229,37 +231,174 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
 		device.VmPath = devPath
 	}
 
-	return updateSpecDeviceList(device, spec)
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.ContainerPath)
 }
 
 // device.Id should be the SCSI address of the disk in the format "scsiID:lunID"
-func virtioSCSIDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox) error {
+func virtioSCSIDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
 	// Retrieve the device path from SCSI address.
 	devPath, err := getSCSIDevPath(s, device.Id)
 	if err != nil {
 		return err
 	}
 	device.VmPath = devPath
 
-	return updateSpecDeviceList(device, spec)
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.ContainerPath)
 }
 
-func nvdimmDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec, s *sandbox) error {
-	return updateSpecDeviceList(device, spec)
+func nvdimmDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.ContainerPath)
 }
 
-// updateSpecDeviceList takes a device description provided by the caller,
-// trying to find it on the guest. Once this device has been identified, the
-// "real" information that can be read from inside the VM is used to update
-// the same device in the list of devices provided through the OCI spec.
-// This is needed to update information about minor/major numbers that cannot
-// be predicted from the caller.
-func updateSpecDeviceList(device pb.Device, spec *pb.Spec) error {
-	// If no ContainerPath is provided, we won't be able to match and
-	// update the device in the OCI spec device list. This is an error.
-	if device.ContainerPath == "" {
+func vfioRebindDevice(bdf string) error {
+	fieldLogger := agentLog.WithField("guest-bdf", bdf)
+	fieldLogger.Debug("VFIO: Attempting to rebind")
+
+	sysPath := filepath.Join(sysBusPrefix, bdf)
+
+	// First we have to wait for the device to appear in the
+	// guest, since the hotplug can take a while.  We can't use
+	// the waiting logic in getDeviceName here, because the udev
+	// devices will only appear as we expect *after* we've
+	// rebound.
+	start := time.Now()
+	for {
+		_, err := os.Stat(sysPath)
+		if err == nil {
+			break
+		}
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("Error accessing %s: %s", sysPath, err)
+		}
+		if elapsed := time.Since(start); elapsed > vfioTimeout {
+			return fmt.Errorf("Timeout waiting for %s (%v)", sysPath, elapsed)
+		}
+		time.Sleep(vfioHotplugPollInterval)
```

Comment:
I added a comment with some times - they're from my laptop, so I don't know how much they'll vary on other systems.  It's about 6.2-6.3s with SHPC, which is slower than I expected - I know there's a 5s delay in the protocol, but I'm surprised that the rest of everything takes a whole extra second.  With PCI-E (hotplug_vfio_on_root_bus=true) it's much faster around 100-120ms.  The latter might suggest reducing that delay granularity.

Adding jitter seems like serious overkill to me: it's hard to imagine multiple containers being in lockstep to a precision that matters with an interval of 20ms.  Plus it's only a stat() in the loop, not anything that hits the kernel particularly hard.

Reply:
If there is any host-side wait on that `stat()` for whatever reason, then your thousand containers will align themselves in lockstep quite naturally.

That `stat()` is not running on bare metal, but in a VM that can get suspended or be in I/O wait for any number of reasons. For all we know, your entire storage might be running off a SAN with 40s worth of I/Os backed up. It finally completes that specific I/O that was blocking your VMs, and then your 1000 containers get woken up at the same time, pile their 1000 stat(), everyone goes waiting again, and we start getting complaints about slow container start time.

Yes, it's probably overkill, but it's also so simple to avoid. Certainly not enough for me to block that PR, but clearly a small improvement that can avoid subtle bugs later.

Reply:
It's a stat() on a guest kernel virtual file system, so, no, a host-side IO delay wouldn't synchronize it.  I really think adding jitter would just be premature complexity.

But... in any case, I'm looking at replacing this path with something looking for a udev event, which will make this moot.

Reply:
uevent based wait now implemented